feat(AIP-136): require GET/POST, use parent instead of singular

toumorokoshi opened the pull request:
The resource singular is a harder pattern to recognize as it
requires a consumer to know the resource singular beforehand. Using
parent aligns it with other collection-oriented methods.

Further explanation for the rationale is in the changes for AIP-136.

fixes #1089

toumorokoshi committed:
feat(AIP-136): require GET/POST, use parent instead of singular

The resource singular is a harder pattern to recognize as it
requires a consumer to know the resource singular beforehand. Using
parent aligns it with other collection-oriented methods.

Further explanation for the rationale is in the changes for AIP-136.

fixes #1089

noahdietz commented on aip/general/0136.md:
```diff
@@ -44,12 +44,10 @@ services. The bullets below apply in all three cases.
-    have no side effects.
-  - Custom methods **should not** use `PATCH` or `DELETE`.
+- The HTTP method **must** be `POST` if the method mutates
+  reosurces or data, and **must** be `GET` otherwise.
```
```suggestion
  resources or data, and **must** be `GET` otherwise.
```

bgrant0607 replied:
or causes some other side effect

noahdietz commented on aip/general/0136.md:
```diff
@@ -44,12 +44,10 @@ services. The bullets below apply in all three cases.
+- The HTTP method **must** be `POST` if the method mutates
+  reosurces or data, and **must** be `GET` otherwise.
+- The parameter for the resource's name **must** be called `name`, and
+  be the only variable in the URI path.
```
Do you think this would be better organized as a "situation" bullet and separate guidance sub-bullets? e.g.

```
- If the method operates on or with resources or data, the following apply:
  - Mutation methods **must** use the HTTP method `POST`
  - Retrieval methods **must** use the HTTP method `GET`
  - The field representing the resource's name **must** be called `name`
  - The `name` field **must** be the only variable in the URI path
```

I think it makes it more digestable, wdyt.

toumorokoshi replied:
Sounds good, let me give that a shot.

noahdietz commented on aip/general/0136.md:
```diff
@@ -119,6 +116,23 @@ An exception to this is for rarely-used, fundamentally imperative operations,
+
+### HTTP path
+
+Similar to standard methods, a custom method should have a `name` or `parent`
```
Can this be worded to sound like a statement of fact rather than guidance? e.g.

> Similar to standard methods, a custom method needs a `name` or `parent` field to indicate the resource that...

bgrant0607 replied:
As per the later comment, it's only a fact in the case that it operates on a resource

toumorokoshi replied:
done.

noahdietz commented on aip/general/0136.md:
```diff
@@ -119,6 +116,23 @@ An exception to this is for rarely-used, fundamentally imperative operations,
+
+### HTTP path
+
+Similar to standard methods, a custom method should have a `name` or `parent`
```
For clarity: How do APIs like Cloud Text-to-Speech & Natural Language fit into this thinking? Most RPCs there don't reference resources at all, pure "data plane" APIs, but they are "custom methods" technically.

Two examples
* [`SynthesizeSpeech`](https://github.com/googleapis/googleapis/blob/9f6e73ef93940ca5df305d5dffcaa3da7b10f0c2/google/cloud/texttospeech/v1/cloud_tts.proto#L53-L56)
* [`AnalyzeSentiment`](https://github.com/googleapis/googleapis/blob/9f6e73ef93940ca5df305d5dffcaa3da7b10f0c2/google/cloud/language/v1/language_service.proto#L37-L40)

bgrant0607 replied:
These should still have virtual parents, like projects and locations, so "parent" still seems relevant. I haven't looked at it deeply, but project override seems less desirable, and regional endpoints don't obviate the need to specify the location in the usual places.

toumorokoshi replied:
I've updated the wording to clarify the guidance only matches if the custom method operates on a resource.

I tend to agree with Brian that there should always be a virtual parent, but I'll probably follow up on that with a more thorough investigation.

bgrant0607 commented on aip/general/0136.md:
```diff
@@ -119,6 +116,23 @@ An exception to this is for rarely-used, fundamentally imperative operations,
+which methods do not mutate data, and which ones do. Methods that only
+read data have first-class concepts in some clients (DataSources in
+Terraform) and clearly indicate to a user which methods can be called
+without riskof runtime impact.
```
risk of

toumorokoshi replied:
thanks! bad word wrapping.

toumorokoshi committed:
address comments

- creating a resource-oriented section to put resource-specific
  guidance.
- making should a must to clarify level of strictness.
- clarifying some wording, more succincet explanations.

toumorokoshi committed:
Merge branch 'master' into toum/custom-methods-get-post

noahdietz commented on aip/general/0136.md:
```diff
@@ -44,25 +44,42 @@ services. The bullets below apply in all three cases.
   verb in the name of the RPC.
   - If word separation is required, `camelCase` **must** be used.
 - The `body` clause in the `google.api.http` annotation **should** be `"*"`.
-  - However, if using `GET` or `DELETE`, the `body` clause **must** be absent.
```
So why don't we allow `DELETE` anymore? Or is that written elsewhere I just haven't seen?

toumorokoshi replied:
That guidance used to be in there:

```
  custom method maps more strongly to another HTTP verb.
  - Custom methods that serve as an alternative to get or list methods (such as
    `Search`) **should** use `GET`. These methods **must** be idempotent and
    have no side effects.
  - Custom methods **should not** use `PATCH` or `DELETE`.
```

But I removed it because "must use GET and POST" was more succinct. And I doubt we want people to add additional custom verbs.

As to why we don't want to use delete? we could, but it just makes it another rarely used category that clients have to handle. Potentially we could add it back, but I was erring toward being conservative and relaxing guidance as needed.

noahdietz replied:
Oh you are right, missed that guidance. Thanks!

bgrant0607 reviewed (approved):
LGTM, thanks

toumorokoshi committed:
Merge branch 'master' into toum/custom-methods-get-post

toumorokoshi committed:
Merge branch 'master' into toum/custom-methods-get-post
